### PR TITLE
[SimplifyCFG] Fold switch over ucmp/scmp to icmp and br

### LINKED ISSUE #105632: [InstCombine] Missed optimization with cmp+switch
Alive2: https://alive2.llvm.org/ce/z/x8aCLE
```
define i8 @src(i32 %a, i32 %b) {
  %res = call i8 @llvm.ucmp.i8.i32(i32 %a, i32 %b)
  switch i8 %res, label %bb1 [
    i8 -1, label %bb2
    i8 0, label %bb2
  ]

bb1:
  ret i8 1

bb2:
  ret i8 2
}

define i8 @tgt(i32 %a, i32 %b) {
  %cmp = icmp ugt i32 %a, %b
  br i1 %cmp, label %bb1, label %bb2

bb1:
  ret i8 1

bb2:
  ret i8 2
}
```

This pattern exists in some rust applications.

cc @Poseydon42 @nikic

### PR DESCRIPTION
**Pull request by nikic**
If we switch over ucmp/scmp and have two switch cases going to the same destination, we can convert into icmp+br.

Fixes https://github.com/llvm/llvm-project/issues/105632.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 1]
```diff
@@ -7131,6 +7131,59 @@ static bool simplifySwitchOfPowersOfTwo(SwitchInst *SI, IRBuilder<> &Builder,
   return true;
 }
 
+/// Fold switch over ucmp/scmp intrinsic to br if two of the switch arms have
+/// the same destination.
+static bool simplifySwitchOfCmpIntrinsic(SwitchInst *SI,
+                                         IRBuilderBase &Builder) {
+  auto *Cmp = dyn_cast<CmpIntrinsic>(SI->getCondition());
+  if (!Cmp || !Cmp->hasOneUse() || SI->getNumCases() != 2)
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7139]: We should handle the following case:
```
  switch i8 %cond, label %default.unreachable [
    i8 -1, label %bb1
    i8 0, label %bb1
    i8 1, label %bb2
  ]
```

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 2]
```diff
@@ -7131,6 +7131,59 @@ static bool simplifySwitchOfPowersOfTwo(SwitchInst *SI, IRBuilder<> &Builder,
...
+    break;
+  }
+  if (Cmp->isSigned())
+    Pred = ICmpInst::getSignedPredicate(Pred);
+
+  // The dominator tree does not change, because it treats multi-edges like
+  // a single edge anyway.
+  Builder.SetInsertPoint(SI->getIterator());
+  Value *ICmp = Builder.CreateICmp(Pred, Cmp->getLHS(), Cmp->getRHS());
+  Builder.CreateCondBr(ICmp, SI->getDefaultDest(), Succ);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7180]: Branch weights should be updated.

**Review by Poseydon42 — APPROVED** [review]
LGTM

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**